hotspot/src/os/linux/vm/cgroupV1Subsystem_linux.hpp

```diff
@@ -0,0 +1,118 @@
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef CGROUP_V1_SUBSYSTEM_LINUX_HPP
+#define CGROUP_V1_SUBSYSTEM_LINUX_HPP
+
+#include "runtime/os.hpp"
+#include "memory/allocation.hpp"
+#include "cgroupSubsystem_linux.hpp"
+
+// Cgroups version 1 specific implementation
+
+class CgroupV1Controller: public CgroupController {
+  private:
+    /* mountinfo contents */
+    char *_root;
+    char *_mount_point;
+
+    /* Constructed subsystem directory */
+    char *_path;
+
+  public:
+    CgroupV1Controller(char *root, char *mountpoint) {
+      _root = os::strdup(root);
+      _mount_point = os::strdup(mountpoint);
+      _path = NULL;
+    }
+
+    virtual void set_subsystem_path(char *cgroup_path);
+    char *subsystem_path() { return _path; }
+};
+
+class CgroupV1MemoryController: public CgroupV1Controller {
+
+  public:
+    bool is_hierarchical() { return _uses_mem_hierarchy; }
+    void set_subsystem_path(char *cgroup_path);
+  private:
+    /* Some container runtimes set limits via cgroup
+     * hierarchy. If set to true consider also memory.stat
+     * file if everything else seems unlimited */
+    bool _uses_mem_hierarchy;
+    jlong uses_mem_hierarchy();
+    void set_hierarchical(bool value) { _uses_mem_hierarchy = value; }
+
+  public:
+    CgroupV1MemoryController(char *root, char *mountpoint) : CgroupV1Controller(root, mountpoint) {
+      _uses_mem_hierarchy = false;
+    }
+
+};
+
+class CgroupV1Subsystem: public CgroupSubsystem {
+
+  public:
+    jlong read_memory_limit_in_bytes();
+    jlong memory_and_swap_limit_in_bytes();
+    jlong memory_soft_limit_in_bytes();
+    jlong memory_usage_in_bytes();
+    jlong memory_max_usage_in_bytes();
+    char * cpu_cpuset_cpus();
+    char * cpu_cpuset_memory_nodes();
+
+    int cpu_quota();
+    int cpu_period();
+
+    int cpu_shares();
+
+    const char * container_type() {
+      return "cgroupv1";
+    }
+    CachingCgroupController * memory_controller() { return _memory; }
+    CachingCgroupController * cpu_controller() { return _cpu; }
+
+  private:
+    julong _unlimited_memory;
+
+    /* controllers */
+    CachingCgroupController* _memory = NULL;
+    CgroupV1Controller* _cpuset = NULL;
+    CachingCgroupController* _cpu = NULL;
+    CgroupV1Controller* _cpuacct = NULL;
```

Comment:
For arm32 platform, if the c++ compiler version being used is earlier than 11, then the non-static member initialization will cause compilation failure.

Reply:
Thanks for pointing this out! I'll investigate how to best address that in this patch series.

Reply:
I've spent some time trying to understand this issue, and this is my current understanding of it. ARM32 is the circumstance which surfaced the problem, but this is not an ARM32 problem. The syntax here was introduced in C++11. The minimal (GCC) compiler version stated for JDK8u is GCC 4.3, which has only partial, experimental support for C++11. You didn't state what compiler you were using but I'm going to assume it's one in the supported list for jdk8u. What I take from this is, JDK8u can't use all C++11 features, since some of the supported compiler-versions don't have all C++11 features; and in particular this feature.

These lines of code are unchanged in jdk mainline at the moment, however, jdk has moved forward the minimal supported compiler version. I doubt that a fix for this would be received favourably in jdk mainline, so we probably need to address it in jdk8u-dev directly (not via backport).

Reply:
The build where I saw compilation failures was on a 32-bit platform with C++ compiler version 4.7.2. Here's the error message: 
error: non-static data member initializers only available with -std=c++11 or -std=gnu++11 [-Werror]

In our jdk8 backport, we initialize these data members using constructor initializer-list.

Reply:
I've just pushed a commit that should resolve this (e6fcc66863e). Thanks for bringing it up!